core/src/Cache/CacheFallbackOperations.cs

```diff
@@ -208,7 +226,56 @@ internal class CacheFallbackOperations
             }
         }
 
-        public static List<MsalRefreshTokenCacheItem> GetAllAdalEntriesForMsal(ILegacyCachePersistance legacyCachePersistance, 
+        private static void RemoveEntriesWithMatchingName(
+            ISet<string> environmentAliases, 
+            string displayableId, 
+            IDictionary<AdalTokenCacheKey, AdalResultWrapper> dictionary)
+        {
+            List<AdalTokenCacheKey> keysToRemove = new List<AdalTokenCacheKey>();
+
+            foreach (KeyValuePair<AdalTokenCacheKey, AdalResultWrapper> kvp in dictionary)
+            {
+                string environment = new Uri(kvp.Key.Authority).Host;
+                string cachedAcccountDisplayableId = kvp.Key.DisplayableId;
+                if (environmentAliases.Contains(environment, StringComparer.OrdinalIgnoreCase) &&
+                    string.Equals(displayableId, cachedAcccountDisplayableId, StringComparison.OrdinalIgnoreCase))
+                {
+                    keysToRemove.Add(kvp.Key);
+                }
+            }
+
+            foreach (AdalTokenCacheKey key in keysToRemove)
+            {
+                dictionary.Remove(key);
+            }
```

Comment:
why not remove directly from the dictionary in the for loop above?

Reply:
I think it's because you can't change the collection which you are enumerating.